Client/Frontend/Browser/PlaylistHelper.swift

```diff
@@ -63,98 +72,137 @@ class PlaylistHelper: NSObject, TabContentScript {
-                                
-                                alert.addAction(UIAlertAction(title: Strings.PlayList.addToPlayListAlertTitle, style: .default, handler: { _ in
-                                    // Update playlist with new items..
-                                    PlaylistItem.addItem(item, cachedData: nil) {
-                                        PlaylistManager.shared.autoDownload(item: item)
-                                        
-                                        log.debug("Playlist Item Added")
-                                        self.delegate?.showPlaylistToast(info: item, itemState: .added)
-                                        UIImpactFeedbackGenerator(style: .medium).bzzt()
-                                    }
-                                }))
-                                alert.addAction(UIAlertAction(title: Strings.cancelButtonTitle, style: .cancel, handler: nil))
-                                self.delegate?.showPlaylistAlert(alert)
-                            }
-                        }
+                    if PlaylistItem.itemExists(item) {
+                        self.updateItem(item)
+                    } else if item.detected {
+                        self.delegate?.openInPlayListActivity(info: nil)
+                        self.delegate?.addToPlayListActivity(info: item, itemDetected: true)
+                        self.delegate?.showPlaylistToast(info: item, itemState: .pendingUserAction)
+                    } else {
+                        self.promptUserForAddingItem(item)
                     }
                 }
             }
         }
     }
+    
+    private func loadAssetPlayability(url: URL, completion: @escaping (Bool) -> Void) {
+        if asset == nil {
+            // We have to create an AVURLAsset here to determine if the item is playable
+            // because otherwise it will add an invalid item to playlist that can't be played.
+            // IE: WebM videos aren't supported so can't be played.
+            // Therefore we shouldn't prompt the user to add to playlist.
+            asset = AVURLAsset(url: url)
+        }
+        
+        let isAssetPlayable = { [weak self]() -> Bool in
+            guard let self = self else { return false }
+            
+            var error: NSError?
+            let status = self.asset?.statusOfValue(forKey: "playable", error: &error)
+            let isPlayable = status == .loaded
+            
+            if let error = error {
+                log.error("Couldn't load asset's playability: \(error)")
+            }
+            return isPlayable
+        }
+        
+        // Performance improvement to check the status first
+        // before attempting to load the playable status
+        if isAssetPlayable() {
+            DispatchQueue.main.async {
+                completion(true)
```

Comment:
Shouldnt we first check `connectionStatus` here? Cause if assetIsPlayable, we do not check the status at all.

Reply:
No because we check the status already. It means the asset was successfully loaded. If the status is NOT playable, then we need to check why: Network maybe, or just unsupported.
If the status is playable, it means the asset was either already loaded previously, or it is just playable already without having to fetch its status or network status. In other words, it isn't reliant on network :) That's the optimization.